### PR TITLE
Seq Extension

### LINKED ISSUE #512: Logging Extension - Seq
### What should we add or change to make your life better?

Create an `extension` to push logs to **Seq**

This would be similar to the existing extension built into Tye:
> Tye can push logs to Elastic stack easily without the need for any SDKs or code changes in your services.

Would be implemented in a method like this into `tye.yaml`:
```
extensions:
  - name: seq
```

### Why is this important to you?

I have implemented Seq into my microservice project, however seeing how simple it was to integrate logging with the built-in Tye method was eye opening.

Would be great to build a lot more extensions like this. E.g. grafana, prometheus, jaeger, 

---

For context -- I have used this library in the past and it made configuring logging very simple with json-like settings to magically determine where to send logs. This could be used as a reference for implementing this feature: 

https://github.com/convey-stack/Convey.Logging

https://github.com/convey-stack/Convey.Logging/blob/master/src/Convey.Logging/Extensions.cs

### PR DESCRIPTION
**Pull request by razfriman**
Attempt to fix #512 

This successfully starts Seq, creates an entry on the Tye dashboard, includes the SEQ logs, enables a link to view the Seq dashboard.

This is my first attempt at contributing to Tye and any feedback or guidance would be highly appreciated.

### CODE REVIEW

**Comment by razfriman** [comment]
Planning to add markdown docs with an example once things are fully functional

**Comment by razfriman** [comment]
Uploading screenshots to include them in the Markdown docs:

Screen Shot 2020-05-29 at 8 37 03 pm

Screen Shot 2020-05-29 at 8 37 12 pm

Screen Shot 2020-05-29 at 8 36 49 pm

**Comment by razfriman** [comment]
**Help Requested:**

Not sure how to get the formatting to pass CI - I am developing on Rider on Mac OSX. Any attempt to run an autoformat fails to pass CI - any tips on this? Also any tips on which encoding to set the file, the default for me is not passing CI.

File: src/Microsoft.Tye.Extensions/Seq/SeqExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            {
+                context.Output.WriteDebugLine("seq service already configured. Skipping...");
+            }
+            else
+            {
+                context.Output.WriteDebugLine("Injecting seq service...");
+
+                var seq = new ContainerServiceBuilder("seq", "datalust/seq")
+                {
+                    EnvironmentVariables =
```
Comment (TimHess) [inline src/Microsoft.Tye.Extensions/Seq/SeqExtensions.cs:24]:
```suggestion
                    EnvironmentVariables =
```

File: src/Microsoft.Tye.Extensions/Seq/SeqExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+                context.Output.WriteDebugLine("seq service already configured. Skipping...");
+            }
+            else
+            {
+                context.Output.WriteDebugLine("Injecting seq service...");
+
+                var seq = new ContainerServiceBuilder("seq", "datalust/seq")
+                {
+                    EnvironmentVariables = 
+                    {
```
Comment (TimHess) [inline src/Microsoft.Tye.Extensions/Seq/SeqExtensions.cs:25]:
```suggestion
                    {
```

**Comment by areller** [comment]
@razfriman I also use Rider. I usually run `dotnet format` (you need to install the tool first)

**Comment by razfriman** [comment]
> @razfriman I also use Rider. I usually run `dotnet format` (you need to install the tool first)

Thanks for the tip, I didn’t know this dotnet tool existed! Will check it out today

File: src/Microsoft.Tye.Extensions/Seq/SeqExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+                }
+            }
+            else if (context.Operation == ExtensionContext.OperationKind.Deploy)
+            {
+                foreach (var project in context.Application.Services.OfType<DotnetProjectServiceBuilder>())
+                {
+                    var sidecar = DiagnosticAgent.GetOrAddSidecar(project);
+
+                    // Use service discovery to find seq
+                    sidecar.Args.Add("--provider:seq=service:seq");
```
Comment (davidfowl) [inline src/Microsoft.Tye.Extensions/Seq/SeqExtensions.cs:81]: Have you tested this?

Comment (razfriman) [reply src/Microsoft.Tye.Extensions/Seq/SeqExtensions.cs:81]: I couldn’t find much information on how this sidecar is implemented. (Maybe the code is in the repository and I totally overlooked it?)

However, I followed the existing patterns from the Elastic extensions and the logs did indeed get pushed into Seq. 

—-

I did test this using the Frontend-backend example and that is how I provided the screenshots. 

I will have a 2nd look to see if I can write up some sort of test to verify more of this functionality

Comment (razfriman) [reply src/Microsoft.Tye.Extensions/Seq/SeqExtensions.cs:81]: This should match up with the provider here: https://github.com/dotnet/tye/blob/master/src/Microsoft.Tye.Hosting.Diagnostics/DiagnosticsProvider.cs#L17

Comment (davidfowl) [reply src/Microsoft.Tye.Extensions/Seq/SeqExtensions.cs:81]: Did you deploy and see if it worked?

Comment (razfriman) [reply src/Microsoft.Tye.Extensions/Seq/SeqExtensions.cs:81]: I actually don't have a Kubernetes cluster set up (locally or cloud). Is there any other way to test to see if this works?

Comment (jkotalik) [reply src/Microsoft.Tye.Extensions/Seq/SeqExtensions.cs:81]: Not today, we don't support other deployment targets nor do we have CI tests which run with kubernetes. We are working on getting some of that bootstrapped soon.

See https://github.com/dotnet/tye/blob/master/docs/tutorials/hello-tye/01_deploy.md#getting-started-with-deployment for some guidance on how to a cluster setup.

File: docs/recipes/logging_seq.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+services:
+- name: backend
+  project: backend/backend.csproj
+- name: frontend
+  project: frontend/frontend.csproj
+```
+
+The `logPath` property here configures the path where Seq will store its data. 
+
+> :bulb: Tye can successfully launch Seq without `logPath`, but ... It's *highly* recommended that you specify a path to store the logs and configuration (add to `.gitignore` if it's part of your repository). Kibana has some mandatory setup the first time you use it, and without persisting the data, you will have to go through it each time.
```
Comment (rynowak) [inline docs/recipes/logging_seq.md:29]: Is this actually true for Seq? it looks like this is a copy-paste from the Elastic stack docs

File: src/Microsoft.Tye.Extensions/Seq/SeqExtensions.cs (hunk 3 above)
Comment (rynowak) [reply src/Microsoft.Tye.Extensions/Seq/SeqExtensions.cs:81]: The code for the sidecar is here if you want to peep it: https://github.com/dotnet/tye/tree/master/src/tye-diag-agent

File: docs/recipes/logging_seq.md (hunk 4 above)
Comment (davidfowl) [reply docs/recipes/logging_seq.md:29]: Yea there's some copy pasta here.

**Comment by jkotalik** [comment]
I'm going to try running these personally 😄

**Comment by jkotalik** [comment]
Found one small issue for tye deploy, otherwise got it working. I'm going to write some docs for `tye run` and `tye deploy` for elastic and seq.

**Comment by jkotalik** [comment]
Thanks @razfriman

**Comment by davidfowl** [comment]
W00p!

**Pull request merged**